### PR TITLE
[CI] Added job adding preview links

### PR DESCRIPTION
**Pull request by mnocon**
Target: master, 4.6, 3.3 and user doc as a follow-up.

Adding a job to automatically generate preview links for all the changed Markdown files, so that we don't have to do this manually. See the comment below.

### Option 1: No Markdown changes in the `docs` directory

Comment saying "No Markdown changes to preview" (https://github.com/ibexa/documentation-developer/pull/2561#issuecomment-2639450610)

### Option 2: Markdown changes
Zrzut ekranu 2025-02-7 o 09 22 17

Bot adds a comment with links to all modified previews (without the diff coloring)

### Option 3: Too many changes

I've chosen an arbitrary limit of 100 files - we can adjust it if needed. Bot adds a comment that too many files were modified to generate a list of links.

### CODE REVIEW

File: .github/workflows/preview_comment.yaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            -   name: Create comment for changed files
+                run: |
+                    file_limit=100
+                    build_url="https://ez-systems-developer-documentation--${{ github.event.pull_request.number }}.com.readthedocs.build/${{inputs.project}}en/${{ github.event.pull_request.number }}/"
+
+                    changed_files=$(git diff --name-only HEAD "origin/$GITHUB_BASE_REF" | grep -E ".md$" || [[ $? == 1 ]])
+                    number_of__changed_files=$(echo "$changed_files" | wc -l)
+
+                    if [[ $changed_files -eq "" ]] ; then
+                      comment="Preview of modified files:: no markdown files changed, no preview needed."
```
Comment (adriendupuis) [inline .github/workflows/preview_comment.yaml:37]: If I modify a code sample, a preview is needed. I would reword to stick to the scope.

```suggestion
                      comment="Preview of modified files:: no Markdown change to preview."
```

File: .github/workflows/preview_comment.yaml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+                      # Guess the URL based on Markdown file location. Remove the .md extension.
+                      urls=$(echo "$changed_files" | cut -d / -f 2- | rev | cut -d '.' -f 2- | rev | sed -e "s ^ $build_url ")
+
+                      left=$(yes "+ [" | head -n "$number_of__changed_files" )
+                      middle=$(yes "](" | head -n "$number_of__changed_files" )
+                      right=$(yes ")\n" | head -n "$number_of__changed_files" )
+
+                      comment=$(paste -d'\0' <(echo "$left") <(echo "$filenames") <(echo "$middle") <(echo "$urls") <(echo "$right"))
+                      comment="Preview of modified files:\n\n$comment"
```
Comment (adriendupuis) [inline .github/workflows/preview_comment.yaml:51]:
```suggestion
                      comment="Preview of modified Markdown files:\n\n$comment"
```

File: .github/workflows/preview_comment.yaml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+                    echo -e $comment > comment.md
+
+            -   name: Find comment
+                id: find-comment
+                uses: peter-evans/find-comment@v3
+                with:
+                    issue-number: ${{ github.event.pull_request.number }}
+                    comment-author: 'github-actions[bot]'
+                    body-includes: 'Preview of modified files'
```
Comment (adriendupuis) [inline .github/workflows/preview_comment.yaml:62]:
```suggestion
                    body-includes: 'Preview of modified Markdown files'
```

**Review by adriendupuis — CHANGES_REQUESTED** [review]

File: .github/workflows/preview_comment.yaml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+              with:
+                  fetch-depth: 0
+
+            -   name: Create comment for changed files
+                run: |
+                    file_limit=100
+                    build_url="https://ez-systems-developer-documentation--${{ github.event.pull_request.number }}.com.readthedocs.build/${{inputs.project}}en/${{ github.event.pull_request.number }}/"
+
+                    changed_files=$(git diff --name-only HEAD "origin/$GITHUB_BASE_REF" | grep -E ".md$" || [[ $? == 1 ]])
+                    number_of__changed_files=$(echo "$changed_files" | wc -l)
```
Comment (adriendupuis) [inline .github/workflows/preview_comment.yaml:34]: I manage to generate the whole link list in one go:
`git diff --name-only HEAD "origin/master" | grep -E "^docs\/.*\.md$" | sed -E "s|^docs/(.*)\.md$|- [\1.md](${build_url}\1/)|"`

Comment (mnocon) [reply .github/workflows/preview_comment.yaml:34]: Thank you, this is like magic 😄 

I've reworked it a bit to avoid harcoding `master` and added a new line as it didn't work in Github Actions (it didn't create a list):
```
                      comment=$(git diff --name-only HEAD "origin/$GITHUB_BASE_REF" | grep -E "^docs\/.*\.md$" | sed -E "s|^docs/(.*)\.md$|- [docs/\1.md](${build_url}\1/)\n|")
```

File: .github/workflows/preview_comment.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+                    echo -e "$comment" > comment.md
+
+            -   name: Find comment
+                id: find-comment
+                uses: peter-evans/find-comment@v3
+                with:
+                    issue-number: ${{ github.event.pull_request.number }}
+                    comment-author: 'github-actions[bot]'
+                    body-includes: 'Preview of modified Markdown files'
```
Comment (adriendupuis) [inline .github/workflows/preview_comment.yaml:50]: I would fix singular/plural issues this way:
```suggestion
                      comment="Preview of modified Markdown: no Markdown change to preview."
                    elif [[ $change_count -gt $file_limit ]] ; then
                      comment="Preview of modified Markdown: Too many files modified in a single PR, preview link list is skipped. ($change_count files &gth; $file_limit)"
                    else
                      comment="Preview of modified Markdown:\n\n$change_list"
                    fi

                    echo -e "$comment" > comment.md

            -   name: Find comment
                id: find-comment
                uses: peter-evans/find-comment@v3
                with:
                    issue-number: ${{ github.event.pull_request.number }}
                    comment-author: 'github-actions[bot]'
                    body-includes: 'Preview of modified Markdown'
```

**Review by adriendupuis — APPROVED** [review]

**Review by julitafalcondusza — APPROVED** [review]
